extensions/interactions/CodeRepl/directives/oppia-interactive-code-repl.component.ts

```diff
@@ -0,0 +1,292 @@
+      this.hasLoaded = true;
+      this.changeDetectionRef.detectChanges();
+    };
+    setTimeout(() => {
+      runAfterTimeout();
+    }, 0);
+  }
+
+  runAndSubmitCode(codeInput: string): void {
+    this.runCode(codeInput, (evaluation, err) => {
+      this.sendResponse(evaluation, err);
+    });
+  }
+  private submitAnswer() {
+    this.runAndSubmitCode(this.code);
+  }
+
+  onEditorChange(code: string): void {
+    this.code = code;
+  }
+
+  runCode(
+      codeInput: string, onFinishRunCallback: (
+        x: string, error: string) => void): void {
+    this.code = codeInput;
+    this.output = '';
+
+    // Evaluate the program asynchronously using Skulpt.
+    Sk.misceval.asyncToPromise(() => {
+      Sk.importMainWithBody('<stdin>', false, codeInput, true);
+    }).then(() => {
+      // Finished evaluating.
+      this.evaluation = '';
+      this.fullError = '';
+
+      if (onFinishRunCallback) {
+        onFinishRunCallback('', '');
+      }
+    }, (err) => {
+      if (!(err instanceof Sk.builtin.TimeLimitError)) {
+        this.evaluation = '';
+        this.fullError = String(err);
+
+        if (onFinishRunCallback) {
+          onFinishRunCallback('', String(err));
+        }
+      }
+    });
+  }
+
+  private initMarkers(editor) {
+    var doc = editor.getDoc();
+
+    // The -1 here is because prepended code ends with a newline.
+    var preCodeNumLines = this.preCode.split('\n').length - 1;
+    var postCodeNumLines = this.postCode.split('\n').length;
+    var fullCodeNumLines = this.code.split('\n').length;
+    var userCodeNumLines = (
+      fullCodeNumLines - preCodeNumLines - postCodeNumLines);
+
+    // Mark pre- and post- code as uneditable, and give it some styling.
+    var markOptions = {
+      atomic: false,
+      readOnly: true,
+      inclusiveLeft: true,
+      inclusiveRight: true
+    };
+
+    if (this.preCode.length !== 0) {
+      doc.markText(
+        {
+          line: 0,
+          ch: 0
+        },
+        {
+          line: preCodeNumLines,
+          ch: 0
+        },
+        angular.extend({}, markOptions, {
+          inclusiveRight: false
+        }));
+
+      for (var i = 0; i < preCodeNumLines; i++) {
+        editor.addLineClass(i, 'text', 'code-repl-noneditable-line');
+      }
+    }
+
+    if (this.postCode.length !== 0) {
+      doc.markText(
+        {
+          line: preCodeNumLines + userCodeNumLines,
+          ch: 0
+        },
+        {
+          line: fullCodeNumLines,
+          ch: 0
+        },
+        markOptions);
+
+      for (var i = 0; i < postCodeNumLines; i++) {
+        editor.addLineClass(
+          preCodeNumLines + userCodeNumLines + i,
+          'text', 'code-repl-noneditable-line');
+      }
+    }
+  }
+
+  sendResponse(evaluation: string, err: string): void {
+    this.currentInteractionService.onSubmit({
+      // Replace tabs with 2 spaces.
+      // TODO(sll): Change the default Python indentation to 4 spaces.
```

Comment:
While we are here can we make an issue for this.

Reply:
Done!